### ISSUE TITLE
Disable the parser's zero-based to one-based arrays transform

### ISSUE THREAD

**Issue by jean-emmanuel** [josdejong/mathjs#782]
Hi, thanks for the great library !

Is there a way to switch from one-based arrays to zero based array in math's parser ?

**Comment by josdejong** [comment]
There is no configuration option for this but you could achieve so by overriding the set of functions that the expression parser uses `math.expression.transform` with the plain JS functions `math`:

```js
// see the one-based behavior of the expression parser
console.log(math.eval('A[1]', {A: [1,2,3]}))   // 1

// see what functions have transforms
console.log(Object.keys(math.expression.transform))
// ["concat", "filter", "forEach", "index", "map", "max", "mean", "min", "range", "subset"]

// replace the object with `transformed` functions for an empty one
math.expression.transform = Object.create(math)

// the expression parser now has zero-based behavior:
console.log(math.eval('A[1]', {A: [1,2,3]}))   // 2
```

One caveat is that when you would import a new function after this change and this function comes with it's own transform, it will be added to `math.expression.transform`, so after an import you have to replace the transform object again.

**Comment by jean-emmanuel** [comment]
Awesome, thanks a lot !

**Comment by josdejong** [comment]
:+1: you're welcome

**Comment by jean-emmanuel** [comment]
Is this still supposed to be true ? Seems like the trick isn't working anymore.

**Comment by josdejong** [comment]
Good point, there have been some internal refactorings. To achieve the same in the latest versions of mathjs, you can do the following:

```js
// see the one-based behavior of the expression parser
console.log(math.eval('A[1]', {A: [1,2,3]}))   // 1

// see what functions have transforms
console.log(Object.keys(math.expression.transform))
// ["concat", "filter", "forEach", "index", "map", "max", "mean", "min", "range", "subset"]

// replace the object containing regular and transformed functions for an object with
// just regular functions. If needed , you can restore orginal behavior afterwards by 
// merging the functions in math.expression.transform again 
// in math.expression.mathWithTransform
math.expression.mathWithTransform = Object.create(math)

// the expression parser now has zero-based behavior:
console.log(math.eval('A[1]', {A: [1,2,3]}))   // 2
```

**Comment by jean-emmanuel** [comment]
Awesome, thank you !

**Comment by wes-r** [comment]
@josdejong :'( Looks like this trick is once again no longer working? Got an update?

**Comment by josdejong** [comment]
Ah, yes indeed that doesn't work anymore since v6 I think. I will see if there is another workaround.

**Comment by josdejong** [comment]
@wes-r this should be easier now in v6, however, I noticed that `IndexNode` currently uses a hard-coded, one-based instance of the `index` function. I've addressed this in b9175a1a29e9b9ac02c899e956551c14c7ccd110 (not yet published).

After this fix is published, it's simply a matter of cherry picking which functions you want to import, in your case that is all functions except transforms:

```js
import { create, all } from 'mathjs'

// default, one-based indices:
const math = create(all)
console.log(math.evaluate('A[1]', { A: [1, 2, 3] })) // 2

// filter transform functions out of the imports:
const allButTransforms = {}
Object.keys(all)
  .filter(key => !key.endsWith('Transform'))
  .forEach(key => {
    allButTransforms[key] = all[key]
  })
// Alternatively using lodash:
//   const allButTransforms = pickBy(all, (value, key) => !key.endsWith('Transform'))

// custom zero-based indices:
const mathWithoutTransforms = create(allButTransforms)
console.log(mathWithoutTransforms.evaluate('A[1]', { A: [1, 2, 3] })) // 1
```
I will let you know as soon as this fix is published.

**Comment by josdejong** [comment]
Ok should be fixed now in `mathjs@6.2.5`, can you give it a try Wesley?

**Comment by wes-r** [comment]
Success! TY!
